src/Juvix/Compiler/Concrete/Translation/FromParsed/Analysis/Scoping.hs

```diff
@@ -928,33 +935,42 @@ checkFunction ::
   Function 'Parsed ->
   Sem r (Function 'Scoped)
 checkFunction Function {..} = do
-  funParameter' <- checkParam
-  let scoped = case funParameter' ^. paramName of
-        Nothing -> id
-        Just s -> withBindLocalVariable (LocalVariable s)
+  funParameters' <- checkParams
+  let scoped =
+        foldr
+          ( \param acc -> case param of
+              Nothing -> acc
+              Just s -> withBindLocalVariable (LocalVariable s) . acc
+          )
+          id
+          (funParameters' ^. paramNames)
   funReturn' <- scoped (checkParseExpressionAtoms _funReturn)
   return
     Function
-      { _funParameter = funParameter',
+      { _funParameters = funParameters',
         _funReturn = funReturn'
       }
   where
-    checkParam :: Sem r (FunctionParameter 'Scoped)
-    checkParam = do
+    checkParams :: Sem r (FunctionParameters 'Scoped)
+    checkParams = do
       paramType' <- checkParseExpressionAtoms _paramType
-      paramName' <- checkParamName
+      paramNames' <- checkParamNames
       return
-        FunctionParameter
-          { _paramName = paramName',
+        FunctionParameters
+          { _paramNames = paramNames',
             _paramImplicit = _paramImplicit,
             _paramType = paramType'
           }
       where
-        FunctionParameter {..} = _funParameter
-        checkParamName :: Sem r (Maybe S.Symbol)
-        checkParamName = case _paramName of
-          Nothing -> return Nothing
-          Just s -> Just <$> freshVariable s
+        FunctionParameters {..} = _funParameters
+        checkParamNames :: Sem r (NonEmpty (Maybe S.Symbol))
+        checkParamNames =
+          mapM
+            ( \case
```

Comment:
personal preference:
In general I don't like lambda functions (or `\case`) unless they are the last argument.
I'd prefer if the lambda was defined in a `let`/`where` or if you use `forM` instead of `mapM`, where the lambda is the last argument

Reply:
I changed it to `forM`